buildroot/share/unit-tests/TestVector.hpp

```diff
@@ -0,0 +1,160 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Copyright (c) 2021 X-Ryl669
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#include <limits.h>
+#include <stdlib.h>
+#include <memory.h>
+#include <new>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <wchar.h>
+// std stuff
+#include <string>
+#include <functional>
+#include <vector>
+
+
+#if defined(__linux__) || defined(__APPLE__)
+  #define _POSIX 1
+#endif
+
+#ifdef _WIN32
+  #include <windows.h>
+  #include <Mmsystem.h>
+  #include <Nb30.h>
+  #include <ws2tcpip.h>
+  #include <tchar.h>
+  #include <iphlpapi.h>
+#elif defined(_POSIX)
+  #include <errno.h>
+  #include <execinfo.h>
+  #include <fcntl.h>
+  #include <netdb.h>
+  #include <netinet/in.h>
+  #include <pthread.h>
+  #include <signal.h>
+  #include <sys/poll.h>
+  #include <sys/socket.h>
+  #include <sys/time.h>
+  #include <sys/types.h>
+  #include <unistd.h>
+  #include <sys/ptrace.h>
+  #include <aio.h>
+  #include <ifaddrs.h>
+
+  #if defined(__linux__)
+    #include <semaphore.h>
+  #elif defined(__APPLE__)
+    #include <sys/sysctl.h>
+    #include <sys/stat.h>
+    #include <sys/uio.h>
+    #include <net/route.h>
+    #include <net/if_dl.h>
+  #endif
+#endif
+
+/** All testing code frameword is declared inside this namespace */
```

Comment:
https://github.com/catchorg/Catch2/ ?

Reply:
Also `frameword`

Reply:
I first thought about using a large testing framework, like Google Test, and Cute. But there are big, and it adds (another) dependency to Marlin and requires also a decent C++ compiler.

In the end, I asked myself what are the real features we needed, and once unchecked the bell and whistle of the *big* player, the result was simple: we don't need much.

So I ended up making a simple 200 lines of code testing framework that does not require to lean a framework, just add you test by copying and modifying an example. 

You can see writing test is very simple (1 line of code), no makefile, no bother to learn. The only "subtle" trick is the `return False` if you want to capture the file & line of code of failure, but a dumb `return false` would work as well.

And for the typo, right, I need to change this.

Reply:
It's simple and nice until you start running into the C++ intricacies, UBs and miscompilations. Of course, one simple framework runs fine on one developer's machine with one version of one compiler.

Gtest, Catch etc have accumulated their size not because they like bloat, but to counter the buzzing minefield that is making C++ constructs compile to something sane _all the time_.